src/libraries/Microsoft.PowerFx.Interpreter/EvalVisitor.cs

```diff
@@ -15,22 +18,35 @@
 
 namespace Microsoft.PowerFx
 {
-    internal class EvalVisitor : IRNodeVisitor<FormulaValue, SymbolContext>
+    // This used ValueTask for async, https://devblogs.microsoft.com/dotnet/understanding-the-whys-whats-and-whens-of-valuetask/ 
+    // Perf comparison of Task vs. ValueTask: https://ladeak.wordpress.com/2019/03/09/valuetask-vs-task 
+    // Use Task for public methods, but ValueTask for internal methods that we expect to be mostly sync. 
+    internal class EvalVisitor : IRNodeVisitor<ValueTask<FormulaValue>, SymbolContext>
     {
         public CultureInfo CultureInfo { get; }
 
-        public EvalVisitor(CultureInfo cultureInfo)
+        private readonly CancellationToken _cancel;
+
+        public EvalVisitor(CultureInfo cultureInfo, CancellationToken cancel)
         {
             CultureInfo = cultureInfo;
+            _cancel = cancel;
+        }
+
+        // Check this cooperatively - especially in any loop. 
```

Comment:
Cooperatively?

Reply:
We're employing a cooperative multitasking approach here - where code is instrumented with checks (ie, the code "cooperates" with us)
[https://en.wikipedia.org/wiki/Cooperative_multitasking](https://en.wikipedia.org/wiki/Cooperative_multitasking)